### PR TITLE
Message Queue

### LINKED ISSUE #2223: Message Queue
Implement a Message Queue service that persistently stores messages outside of task sessions.

It would simplify `TASK_CONN_TYPES`. Developer would no longer need to create infamous `_add_pending_request`, `__connection_established_for_XXX`, `__connection_failure_for_XXX`, `__connection_final_failure_for_XXX` quartet. He would instead just call `message_queue.put(node_id, ..., msg description)`

It would also solve the persistence issue.

_Updated 2018-02-26: more elaborated gains @jiivan_
_Updated 2018-02-26: no longer needed by Concent -- @jiivan_

* [x] make `TaskSession` stateless
* [x] implement the Message Queue service
* [x] implement the Message Queue persistence layer
* [x] add unit tests for the new Message Queue service
* [x] simplify `TaskServer` - replace the connection types with a unified transport layer utilizing the Message Queue
* [x] Use MQ in offer_chosen callback
* [x] Implement task session timeouts

### LINKED ISSUE #2403: Don't duplicate information in send_report_computed_task
Information that are passed to this function as arguments and the one that are packed in node_info are redundant.

### LINKED ISSUE #2404: Replace TASK_CONN_TYPES representation from int labels to strings

### LINKED ISSUE #4117: AttributeError: 'TaskServer' object has no attribute 'key_id'
## Description

**Golem Version**: GOLEM Version: 0.19.1+dev277.g7435862

**Golem-Messages version** (leave empty if unsure): golem_messages Version: 3.3.0

**Electron version** (if used): 0.1.19

**OS** [e.g. Windows 10 Pro]: Windows 10 Pro

**Branch** (if launched from source): develop

**Mainnet/Testnet**: mainnet

**Description of the issue**:

after task is added on requestor's node on providing nodes error appears in logs

**Actual result**:

Despite critical, subtask has been computed, but node is not providing again for the requestor. Nodes are not banned for each other.

**Screenshots**:

_If applicable, add screenshots to help explain your problem._

## Steps To Reproduce
_Short description of steps to reproduce the behavior:_
e.g.
1. Run Golem from source on develop branch
2. Add short task on requesting node e.g. Helicopter 400x400 1 frame, 3 subtasks
3. Check provider's logs for critical

## Logs and any additional context

```
CRITICAL [twisted                            ]
Traceback (most recent call last):
  File "C:\Users\Magdalena Szydlowska\Projects\golem-env\lib\site-packages\twisted\internet\defer.py", line 653, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "C:\Users\Magdalena Szydlowska\Projects\golem-env\lib\site-packages\twisted\internet\task.py", line 865, in <lambda>
    d.addCallback(lambda ignored: callable(*args, **kw))
  File "C:\Users\Magdalena Szydlowska\Projects\golem\golem\task\server\resources.py", line 224, in <lambda>
    lambda *_: self._handshake_timeout(self.key_id)
AttributeError: 'TaskServer' object has no attribute 'key_id'
```

## Proposed Solution?
_(Optional: What could be a solution for that issue)_

### PR DESCRIPTION
**Pull request by jiivan**
Final step of implementing **message queue** in golem. Plug in all functionality in **TaskServer**. Move unnecessary logic from **TaskSession** into **TaskServer** or it's submodules in `golem.task.server`.

fixes: #2223 
fixes: #2403 
fixes: #4117 
fixes: #2404

### CODE REVIEW

File: golem/task/server/helpers.py [hunk 1]
```diff
@@ -80,3 +90,131 @@ def on_error(exc, *_args, **_kwargs):
...
+            common.node_info_str(
+                waiting_task_result.owner.node_name,
+                waiting_task_result.owner.key,
+            ),
+            waiting_task_result.task_id,
+            waiting_task_result.subtask_id,
+        )
+        return
+
+    i_am: LocalNode = task_server.node
```
Comment (shadeofblue) [inline golem/task/server/helpers.py]: maybe `me` or `my_node` would be better, grammatically? ;)

File: golem/task/server/helpers.py [hunk 2]
```diff
@@ -80,3 +90,131 @@ def on_error(exc, *_args, **_kwargs):
         client_options=client_options,
         output_dir=output_dir
     )
+
+def send_report_computed_task(task_server, waiting_task_result) -> None:
+    """ Send task results after finished computations
+    """
+    from golem.task.tasksession import copy_and_sign
```
Comment (shadeofblue) [inline golem/task/server/helpers.py]: instead of importing inline, please move this function out of `tasksession.py` ... it's not really an integral part of `tasksession` and, in fact, could actually be included in the `message.base.Message` class...

File: golem/network/transport/msg_queue.py [hunk 3]
```diff
@@ -5,22 +5,19 @@
...
-    from golem_messages import message
-
-
 logger = logging.getLogger(__name__)
 READ_LOCK = threading.Lock()
 
 
-def put(node_id: str, msg: 'message.base.Base') -> None:
+def put(node_id: str, msg: message.base.Message) -> None:
+    assert not isinstance(msg, message.base.Disconnect)
```
Comment (shadeofblue) [inline golem/network/transport/msg_queue.py]: maybe add some assertion failure message?

File: golem/task/server/helpers.py [hunk 4]
```diff
@@ -80,3 +90,131 @@ def on_error(exc, *_args, **_kwargs):
...
+    task_to_compute = history.get(
+        message_class_name='TaskToCompute',
+        node_id=waiting_task_result.owner.key,
+        task_id=waiting_task_result.task_id,
+        subtask_id=waiting_task_result.subtask_id
+    )
+
+    if not task_to_compute:
+        logger.warning(
+            "I won't report computed task. TTC missing."
```
Comment (shadeofblue) [inline golem/task/server/helpers.py]: I think expressing it as `Cannot send ReportComputedTask - TTC missing` would be potentially less ambiguous

File: golem/task/server/helpers.py [hunk 5]
```diff
@@ -80,3 +90,131 @@ def on_error(exc, *_args, **_kwargs):
...
+    task_to_compute = history.get(
+        message_class_name='TaskToCompute',
+        node_id=waiting_task_failure.owner.key,
+        task_id=waiting_task_failure.task_id,
+        subtask_id=waiting_task_failure.subtask_id
+    )
+
+    if not task_to_compute:
+        logger.warning(
+            "I won't report task failure. TTC missing."
```
Comment (shadeofblue) [inline golem/task/server/helpers.py]: same here ... maybe `Cannot send TaskFailure` would be clearer?

File: golem/task/server/queue_.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
+import logging
```
Comment (shadeofblue) [inline golem/task/server/queue_.py:1]: @jiivan I'm unsure I understand the file rename to `queue_` -> is that some idiom?

File: golem/task/server/queue_.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+logger = logging.getLogger(__name__)
+
+class TaskMessagesQueueMixin:
+    """Message Queue functionality for TaskServer"""
+
+    task_keeper: 'taskkeeper.TaskHeaderKeeper'
+    forwarded_session_requests: typing.Dict[str, dict]
+
+    def __init__(self):
+        # None - PendingConnection
```
Comment (shadeofblue) [inline golem/task/server/queue_.py]: it's not clear what these comments refer to...

File: golem/task/server/queue_.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            return
+        self.remove_session(session)
+
+    def remove_session(self, session):
+        session.disconnect(
+            message.base.Disconnect.REASON.NoMoreMessages,
+        )
+        self.remove_pending_conn(session.conn_id)
+        if session.task_computer is not None:
+            # XXX ???
```
Comment (shadeofblue) [inline golem/task/server/queue_.py]: ???

File: golem/task/server/queue_.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+    def msg_queue_connection_established(
+            self,
+            session: 'TaskSession',
+            conn_id,
+            node_id,
+    ):
+        try:
+            if self.sessions[node_id] is not None:
+                # There is a session already established
```
Comment (shadeofblue) [inline golem/task/server/queue_.py:117]: meaning, for this node_id, right? in which case all messages will be processed within that other session, right? it would be good if the comment was clearer about that... 

iow, please eli5 a bit more, thank you :)

File: golem/task/server/queue_.py [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+            pc.time = time.time()
+
+    def msg_queue_connection_final_failure(
+            self,
+            conn_id,
+            node_id,
+            *_args,
+            **_kwargs,
+    ):
+        self.remove_pending_conn(conn_id)
```
Comment (shadeofblue) [inline golem/task/server/queue_.py:157]: maybe add some debug log?

**Review by shadeofblue — APPROVED** [review]
looks good as far as I could tell - would be good to have it tested by the QA before we merge it, adding to the QA project's backlog

File: golem/task/server/queue_.py [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+            self,
+            session: 'TaskSession',
+            conn_id,
+            node_id,
+    ):
+        try:
+            if self.sessions[node_id] is not None:
+                # There is a session already established
+                session.dropped()
+                return
```
Comment (maaktweluit) [inline golem/task/server/queue_.py:121]: I think in the old code it also send a duplicate peer message here

File: golem/task/server/verification.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+        :param str subtask_id: subtask that has wrong result
+        :param SubtaskResultsRejected.Reason reason: the rejection reason
+        """
+
+        from golem.task.tasksession import copy_and_sign
+
+        logger.debug(
+            'send_result_rejected. reason=%r, rct=%r',
+            reason,
```
Comment (maaktweluit) [inline golem/task/server/verification.py:117]: you can use `reason.name` for a cleaner log

File: golem/task/taskserver.py [hunk 13]
```diff
@@ -409,17 +401,27 @@ def send_task_failed(
...
 
     def new_connection(self, session):
-        if self.active:
-            self.task_sessions_incoming.add(session)
-        else:
+        if not self.active:
             session.disconnect(message.base.Disconnect.REASON.NoMoreMessages)
+            return
+        logger.debug(
+            'Incoming TaskSession. host=%r, port=%r',
```
Comment (maaktweluit) [inline golem/task/taskserver.py]: `address=%s:%d` makes it easier to find across logs ( when used consistently )

File: golem/task/taskserver.py [hunk 14]
```diff
@@ -1280,14 +919,3 @@ def __init__(self, task_id, subtask_id, err_msg, owner):
         self.subtask_id = subtask_id
         self.owner = owner
         self.err_msg = err_msg
-
-
-# TODO: Get rid of archaic int labels and use plain strings instead. issue #2404
```
Comment (maaktweluit) [inline golem/task/taskserver.py:1285]: Does this also close #2404 ?

**Review by maaktweluit — COMMENTED** [review]
LGTM!

left some small comments

will check it again tomorrow and approve :)

File: golem/task/taskserver.py (hunk 14 above)
Comment (jiivan) [reply golem/task/taskserver.py:1285]: Yes, added it to the list.

File: golem/task/server/queue_.py (hunk 6 above)
Comment (jiivan) [reply golem/task/server/queue_.py:1]: It's related to a bug in linters (and possibly pytest). They try to import built-in [queue](https://docs.python.org/3/library/queue.html) like it's still python2...

File: golem/task/server/queue_.py (hunk 11 above)
Comment (jiivan) [reply golem/task/server/queue_.py:121]: It's in a peersession:
https://github.com/golemfactory/golem/blob/develop/golem/network/p2p/peersession.py#L523-L535

I'm not sure if it's needed in TaskSession context.

File: golem/task/server/verification.py (hunk 12 above)
Comment (jiivan) [reply golem/task/server/verification.py:117]: Only if `reason` is never `None`...

File: golem/task/server/queue_.py (hunk 11 above)
Comment (maaktweluit) [reply golem/task/server/queue_.py:121]: ahh, indeed, sorry i mixed those up

File: golem/task/server/verification.py (hunk 12 above)
Comment (maaktweluit) [reply golem/task/server/verification.py:117]: :(

**Comment by ederenn** [comment]
Tested on:
```0.19.1+dev358.ge973167
INFO     [golemapp                           ] Protocol Version: 32
INFO     [golemapp                           ] golem_messages Version: 3.3.1
```


After connecting to other nodes in network I've received 

encontered on Linux18 and Windows 10
```
INFO     [golem.client                       ] Resumed
Unhandled Error
Traceback (most recent call last):
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\python\log.py", line 103, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\python\log.py", line 86, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\python\context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\python\context.py", line 85, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\internet\iocpreactor\reactor.py", line 119, in _callEventCallback
    evt.callback(rc, numBytes, evt)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\internet\iocpreactor\tcp.py", line 552, in cbAccept
    self.handleAccept(rc, evt)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\internet\iocpreactor\tcp.py", line 594, in handleAccept
    protocol.makeConnection(transport)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\internet\protocol.py", line 510, in makeConnection
    self.connectionMade()
  File "C:\Users\ederenn\Projects\golem\golem\network\transport\tcpnetwork.py", line 435, in connectionMade
    self.server.new_connection(self.session)
  File "C:\Users\ederenn\Projects\golem\golem\task\taskserver.py", line 405, in new_connection
    session.disconnect(message.base.Disconnect.REASON.NoMoreMessages)
  File "C:\Users\ederenn\Projects\golem\golem\network\transport\session.py", line 108, in disconnect
    self._send_disconnect(reason)
  File "C:\Users\ederenn\Projects\golem\golem\network\transport\session.py", line 140, in _send_disconnect
    self.send(message.base.Disconnect(reason=reason))
  File "C:\Users\ederenn\Projects\golem\golem\task\tasksession.py", line 897, in send
    raise RuntimeError('Connection unverified')
builtins.RuntimeError: Connection unverified

CRITICAL [twisted                            ] Unhandled Error
Traceback (most recent call last):
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\python\log.py", line 103, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\python\log.py", line 86, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\python\context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\python\context.py", line 85, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\internet\iocpreactor\reactor.py", line 119, in _callEventCallback
    evt.callback(rc, numBytes, evt)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\internet\iocpreactor\tcp.py", line 552, in cbAccept
    self.handleAccept(rc, evt)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\internet\iocpreactor\tcp.py", line 594, in handleAccept
    protocol.makeConnection(transport)
  File "C:\Users\ederenn\Projects\golem-env\lib\site-packages\twisted\internet\protocol.py", line 510, in makeConnection
    self.connectionMade()
  File "C:\Users\ederenn\Projects\golem\golem\network\transport\tcpnetwork.py", line 435, in connectionMade
    self.server.new_connection(self.session)
  File "C:\Users\ederenn\Projects\golem\golem\task\taskserver.py", line 405, in new_connection
    session.disconnect(message.base.Disconnect.REASON.NoMoreMessages)
  File "C:\Users\ederenn\Projects\golem\golem\network\transport\session.py", line 108, in disconnect
    self._send_disconnect(reason)
  File "C:\Users\ederenn\Projects\golem\golem\network\transport\session.py", line 140, in _send_disconnect
    self.send(message.base.Disconnect(reason=reason))
  File "C:\Users\ederenn\Projects\golem\golem\task\tasksession.py", line 897, in send
    raise RuntimeError('Connection unverified')
builtins.RuntimeError: Connection unverified
```

network consists of 5 nodes, 1 requestor Linux, 4 providers. Linux, Mac OS, Windows
At first try, 3 different providers were computing for requestor. Only Mac and Linux provided with success, providing by Windows ended  at every try in failure.
task, 1 frame 6 subtasks, was finished successfully but mainly on one node. two providers computed only one subtask in task the rest has been finished on third provider.
Situation has repeated in next four tasks. 
After restart of a provider providing nodes were not staring computations at all, no additional errors in regular logs. Might be worth trying on debug.

After another restart of a requestor nodes started to provide. 

Errors encountered during first round of computations:

from requestor:
```
2019-04-18 15:42:19 INFO     golem.task.taskserver               Computation for task '3c98b2c8-61df-11e9-8e9d-5ad86f3605c2' failed: 'Error downloading resources: [Failure instance: Traceback (failure with no frames): <class \'requests.exceptions.HTTPError\'>: {"error":"d620bddd8cf0439914e84f631bcabead1849a6a8572ca3ac866d398bf53fa5e5 download timed out after 234 s"}\n]'.
``` 
 subtask has been provided by Windows node

when subtask checked from cli:
```
>> subtasks show 3c98b2c8-61df-11e9-8e9d-5ad86f3605c2
node_id: d178f6b5bb2c4533ef16df75f8407a99c8d10f1475639d76ac94dfc272d814baf42ebffac96c0a14e408f8bb882043045ffedd3a43d07689e7916fa0042a2b93
node_name: vm verif
progress: 1.0
results: []
status: Failure
stderr: Error downloading resources: [Failure instance: Traceback (failure with no frames): <class 'requests.exceptions.HTTPError'>: {"error":"d620bddd8cf0439914e84f631bcabead1849a6a8572ca3ac866d398bf53fa5e5 download timed out after 234 s"}
]
stdout: 
subtask_id: 3c98b2c8-61df-11e9-8e9d-5ad86f3605c2
time_remaining: 0.0
time_started: 1555594703.4309285
```


from requestor:
```
2019-04-18 15:39:33 CRITICAL twisted                             Unhandled error in Deferred:
2019-04-18 15:39:33 CRITICAL twisted                             
Traceback (most recent call last):
  File "/home/ederenn/projects/golem-env/lib/python3.6/site-packages/twisted/internet/defer.py", line 653, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/ederenn/projects/golem/golem/task/tasksession.py", line 379, in _offer_chosen
    ctd["resources"] = self.task_server.get_resources(msg.task_id)
TypeError: 'NoneType' object does not support item assignment
```

from requestor:
```
2019-04-18 15:39:33 INFO     golem.task.taskmanager              no more computation needed: 2e2d79ae-61df-11e9-abf1-5ad86f3605c2
2019-04-18 15:39:33 CRITICAL twisted                             Unhandled error in Deferred:
2019-04-18 15:39:33 CRITICAL twisted                             
Traceback (most recent call last):
  File "/home/ederenn/projects/golem-env/lib/python3.6/site-packages/twisted/internet/defer.py", line 653, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/ederenn/projects/golem/golem/task/tasksession.py", line 379, in _offer_chosen
    ctd["resources"] = self.task_server.get_resources(msg.task_id)
TypeError: 'NoneType' object does not support item assignment
2019-04-18 15:39:46 INFO     golem.task.tasksession              Received offer to compute. task_id='2e2d79ae-61df-11e9-abf1-5ad86f3605c2', node="'Oncoming Storm'(6c15fa96..d6a66c61)"
```




from provider:
```
019-04-18 14:37:36 ERROR    golem.task.taskkeeper               Problem restoring dumpfile: /home/ederenn/.local/share/golem/default/mainnet/tasks/comp_task_keeper.pickle; deleting broken file
Traceback (most recent call last):
  File "/home/ederenn/projects/golem/golem/task/taskkeeper.py", line 160, in restore
    data = pickle.load(f)
AttributeError: 'TaskHeader' object has no attribute 'resource_size'
```


repeated on providers several times in row 
```
2019-04-18 16:01:57 ERROR    twisted                             TypeError: __init__() missing 1 required keyword-only argument: 'outfilebasename': Traceback (most recent call last):
  File "/home/ederenn/projects/golem-env/lib/python3.6/site-packages/autobahn/twisted/websocket.py", line 162, in _onMessage
    self.onMessage(payload, isBinary)
  File "/home/ederenn/projects/golem-env/lib/python3.6/site-packages/autobahn/wamp/websocket.py", line 95, in onMessage
    self._session.onMessage(msg)
  File "/home/ederenn/projects/golem-env/lib/python3.6/site-packages/autobahn/wamp/protocol.py", line 895, in onMessage
    on_reply = txaio.as_future(endpoint.fn, *invoke_args, **invoke_kwargs)
  File "/home/ederenn/projects/golem-env/lib/python3.6/site-packages/txaio/tx.py", line 417, in as_future
    return maybeDeferred(fun, *args, **kwargs)
--- <exception caught here> ---
  File "/home/ederenn/projects/golem-env/lib/python3.6/site-packages/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File "/home/ederenn/projects/golem/golem/client.py", line 901, in get_task_stats
    'provider_state': self.get_provider_status(),
  File "/home/ederenn/projects/golem/golem/client.py", line 1355, in get_provider_status
    task_computer.get_progress()
  File "/home/ederenn/projects/golem/golem/task/taskcomputer.py", line 219, in get_progress
    **c.extra_data,
builtins.TypeError: __init__() missing 1 required keyword-only argument: 'outfilebasename'
```

**Comment by jiivan** [comment]
1. `builtins.RuntimeError: Connection unverified` resolved in 64bebf8
1. CTD None (NoMoreSubtasks before offer_chosen) resolved in cd15c3d
1. `AttributeError: 'TaskHeader' object has no attribute 'resource_size'` looks unrelated to this PR
1. `builtins.TypeError: __init__() missing 1 required keyword-only argument: 'outfilebasename'` also looks unrelated
